skbeam/core/accumulators/projector.py

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env python
+
+from __future__ import division
+import numpy as np
+
+"""
+authors: Mikhail Dubrovin, TJ Lane, Christopher O'Grady
+"""
+
+
+class Projector(object):
+
+    def __init__(self, bin_values, nbins, weights=None, norm=True):
+        """
+        Parameters
+        ----------
+        bin_values : np.ndarray
+            For each pixel, this is the value of that determines which bin
+            a pixel's intensity will contribute to
+        weights : np.ndarray
+            A weight for each pixel with the same shape as bin_values (set
+            to zero to ignore a pixel)
+        nbins : int
+            The number of bins to project into
+        """
+
+        self.bin_values = bin_values
+        self.weights = weights
+        self.nbins = nbins
+        self.norm = norm
+
+        # _ibv = included bin values
+        # this is used to select only unmasked pixels in the analysis
+        # (i.e. those with weight!=0)
+        if weights is not None:
+            self._ibv = (weights != 0)
+        else:
+            self._ibv = np.s_[:]  # slice that returns arr unchanged
+
+        bv = self.bin_values[self._ibv]
+        wt = self.weights[self._ibv]
+
+        binvalRange = bv.max() - bv.min()
+        # add a little bit to not have the bv.max() value create an extra bin
+        # in the np.floor statement below
+        binvalRange *= 1.0+1e-9
+        self.bin_width = binvalRange / (float(self.nbins))
+
+        self._bin_assignments = np.floor((bv - bv.min()) /
+                                         self.bin_width).astype(np.int32)
+        smallval = np.finfo(np.float).eps
+        if weights is None:
+            self._normalization_array = \
+                (np.bincount(self._bin_assignments.flatten()) + smallval) \
```

Comment:
it looks like `smallval` is being used here to prevent divide by 0 warnings later?  It is probably better to replace the 0s by `np.nan` rather than add a small value to every normalization